### PR TITLE
[Python] f-string with const propagation

### PR DESCRIPTION
**Pull request by ulziibay**
Adds tests for python f-string with constant propagation. This clarifies  and closes https://github.com/returntocorp/sgrep/issues/326

Testing: 
`make test`

### CODE REVIEW

File: sgrep/tests/python/equivalence_f_string.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  ww = "bar"
+  query = f"ASD{ww}ASASD"
+
+def foo3():
+  # OK:
+  www = "bar"
+  query = f"SELECT {www}"
+
+def foo4():
+  # OK:
```
Comment (nbrahms) [inline sgrep/tests/python/equivalence_f_string.py]: Shouldn't this match your pattern?

Comment (ulziibay) [reply sgrep/tests/python/equivalence_f_string.py]: `...` is matching more than 0 characters. Something like
```
$M = "..."
$N = "..."
...
$Q = f"...{$M}...{$N}"
``` matches though!

Comment (nbrahms) [reply sgrep/tests/python/equivalence_f_string.py]: Hmm, I'd think constant propagation should treat these as equivalent. Maybe a bug to file?

Comment (nbrahms) [reply sgrep/tests/python/equivalence_f_string.py]: Honestly I think `...` should match 0 characters.

Comment (aryx) [reply sgrep/tests/python/equivalence_f_string.py]: "..." matches 0 characters.
The issue here is that f"...{$M}"..." is internally rewritten as
special_concat("...", $M, "...") 
and a string like f"foo{a}" is internally rewritten as
special_concat("foo", a)
and the issue here is that the number of arguments are different so the engine will say "no match".

We need to handle specially calls to special_concat and accept that "..." matches not only a string, or an empty string, or nothing at all.
In fact we probably want f"...{$X}..." to match any interpolated expression, not just the first one, so you also don't have to write another pattern when an f-string use 2 interpolated expressions as in f"one {a} other stuff {b}"

**Comment by aryx** [comment]
BTW ulzii, you can use hub pull-request -b develop
when you create the pull request, so that's it based on develop instead of master
(each time you submit a PR there is a huge list of file modifications, probably because you don't use the good "base").

**Comment by ulziibay** [comment]
@aryx good to know about the -b flag

File: sgrep/tests/python/equivalence_f_string.py (hunk 1 above)
Comment (ulziibay) [reply sgrep/tests/python/equivalence_f_string.py]: @aryx: I have update the generic matcher such that  "..." could match many more arguments, thus taking care of the case  where  the number of arguments are different.
However, I use `constant_propagation_and_evaluate_literal` to determine if variable `www` is a literal.  As the moment, the constant propagation in python does not seem to keep track of `www` as a literal

File: sgrep/lib/generic_vs_generic.ml [hunk 2]
```diff
@@ -1033,6 +1032,21 @@ and m_list__m_argument (xsa: A.argument list) (xsb: A.argument list) =
       (* can match more *)
       (m_list__m_argument ((A.Arg (A.Ellipsis i))::xsa) xsb)
 
+  (* dots: '...', can also match no argument *)
```
Comment (aryx) [inline sgrep/lib/generic_vs_generic.ml]: maybe add a comment here saying interpolated strings are transformed in Call(Spedial(Concat, ...)
and we want patterns like f"...{$X}...", which are expanded to Call(Special(Concat, [L"..."; Id "$X"; L"..."])) to
match concrete code like f"foo{a}"

Comment (ulziibay) [reply sgrep/lib/generic_vs_generic.ml]: Ok

File: sgrep/tests/python/equivalence_f_string.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  # ERROR:
+  www = "bar"
+  query = f"SELECT {www}"
+
+def foo4():
+  # TODO:
+  # OK:
+  ww = "foo"
+  www = "bar"
+  query = f"SELECT {www} and {ww}"
```
Comment (nbrahms) [inline sgrep/tests/python/equivalence_f_string.py]: I'm also confused why this wouldn't match?

File: sgrep/tests/python/equivalence_f_string_2.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  select = "select * "
+  name = "foo"
+  query = f"{select} from foo " + f"where name={name}"
+
+def foo2a():
+  # TODO:
+  # OK:
+  select = "select * "
+  name = "foo"
+  query = f"{select} from foo where name={name}"
```
Comment (nbrahms) [inline sgrep/tests/python/equivalence_f_string_2.py]: Also it seems like this should match

Comment (nbrahms) [reply sgrep/tests/python/equivalence_f_string_2.py]: NM I'm silly

File: sgrep/tests/python/equivalence_f_string.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  ww = "bar"
+  query = f"ASD{ww}"
+
+def foo3():
+  # ERROR:
+  www = "bar"
+  query = f"SELECT {www}"
+
+def foo4():
+  # TODO:
```
[the remark this answers is not in the document]
Comment (ulziibay) [reply sgrep/tests/python/equivalence_f_string.py]: From https://github.com/returntocorp/sgrep/pull/347#discussion_r401177492, 
I use constant_propagation_and_evaluate_literal to determine if variable www is a literal. As the moment, the constant propagation in python does not seem to keep track of www as a literal

File: sgrep/tests/python/equivalence_f_string_2.py (hunk 4 above)
Comment (ulziibay) [reply sgrep/tests/python/equivalence_f_string_2.py]: From https://github.com/returntocorp/sgrep/pull/347#discussion_r401177492,  I use constant_propagation_and_evaluate_literal to determine if variable name is a literal. As the moment, the constant propagation in python does not seem to keep track of name as a literal

File: sgrep/tests/python/equivalence_f_string.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ulziibay) [reply sgrep/tests/python/equivalence_f_string.py]: Yes we still need TODO: From #347 (comment), I use constant_propagation_and_evaluate_literal to determine if variable name is a literal. As the moment, the constant propagation in python does not seem to keep track of name as a literal

**Comment by brendongo** [comment]
For reference this is working here: https://sgrep.live/eG8

**Pull request merged**